troi/listenbrainz/listens.py

```diff
@@ -58,15 +46,44 @@ def _fetch_recent_listens_index(self):
 
             for listen in data["listens"]:
                 listened_at = listen["listened_at"]
-                mbid = self.get_recording_mbid(listen)
 
-                if mbid:
-                    index[mbid] = max(index[mbid], listened_at)
+                # add both user submitted mbid and mapped mbid to index
+
+                additional_info = listen["track_metadata"]["additional_info"]
+                user_submitted_mbid = additional_info.get("recording_mbid")
+                if user_submitted_mbid:
+                    index[user_submitted_mbid] = max(index[user_submitted_mbid], listened_at)
+
+                mbid_mapping = listen["track_metadata"].get("mbid_mapping")
+                if mbid_mapping:
+                    mapped_mbid = mbid_mapping.get("recording_mbid")
+                    index[mapped_mbid] = max(index[mapped_mbid], listened_at)
 
             min_ts = data["listens"][0]["listened_at"]
 
         return index
 
+    def _get_latest_listened_ts(self, r: Recording):
+        # check latest listened timestamp for original mbid
+        ts1 = self.index.get(r.mbid)
+
+        # check latest listened timstamp for canonical mbid
+        ts2 = None
+        if r.listenbrainz is not None and r.listenbrainz.get("canonical_recording_mbid"):
+            canonical_mbid = r.listenbrainz.get("canonical_recording_mbid")
+            ts2 = self.index.get(canonical_mbid)
+
+        # if we have timestamp for both canonical and normal mbid take the maximum of two else whichever
+        # one is available. if none is available continue ahead
```

Comment:
what was your reasoning for choosing the larger one?

Reply:
Larger one means that it was listened more recently. Say one mbid listened 3 days ago and the other mbid listened yesterday. Store yesterday's timestamp for use in filter later.